engine/pkg/client/discovery_service.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client
+
+import (
+	"context"
+
+	"github.com/pingcap/log"
+	"github.com/pingcap/tiflow/pkg/retry"
+	"go.uber.org/zap"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tiflow/engine/enginepb"
+	"github.com/pingcap/tiflow/engine/model"
+	"github.com/pingcap/tiflow/engine/pkg/client/internal"
+)
+
+// DiscoveryClient is a client to the Discovery service on the server master.
+type DiscoveryClient interface {
+	// RegisterExecutor registers an executor. The server
+	// will allocate and records a UUID.
+	RegisterExecutor(
+		ctx context.Context,
+		request *enginepb.RegisterExecutorRequest,
+	) (model.ExecutorID, error)
+
+	// Heartbeat sends a heartbeat message to the server.
+	Heartbeat(
+		ctx context.Context,
+		request *enginepb.HeartbeatRequest,
+	) (*enginepb.HeartbeatResponse, error)
+
+	// RegisterMetaStore registers a new metastore.
+	// TODO discuss whether this method can be moved to another service.
+	RegisterMetaStore(
+		ctx context.Context,
+		request *enginepb.RegisterMetaStoreRequest,
+	) error
+
+	// QueryMetaStore queries the details of a metastore.
+	// TODO discuss whether this method can be moved to another service.
+	QueryMetaStore(
+		ctx context.Context,
+		request *enginepb.QueryMetaStoreRequest,
+	) (*enginepb.QueryMetaStoreResponse, error)
+}
+
+var _ DiscoveryClient = &discoveryClient{}
+
+type discoveryClient struct {
+	cli enginepb.DiscoveryClient
+}
+
+// NewDiscoveryClient returns a DiscoveryClient.
+func NewDiscoveryClient(cli enginepb.DiscoveryClient) DiscoveryClient {
+	return &discoveryClient{cli: cli}
+}
+
+func (c *discoveryClient) RegisterExecutor(
+	ctx context.Context,
+	request *enginepb.RegisterExecutorRequest,
+) (model.ExecutorID, error) {
+	var ret model.ExecutorID
+	err := retry.Do(ctx, func() error {
+		call := internal.NewCall(
+			c.cli.RegisterExecutor,
+			request,
+			// RegisterExecutor is not idempotent in general
+			// TODO review idempotency
+			// internal.WithForceNoRetry()
+		)
+		resp, err := call.Do(ctx)
+		if err != nil {
+			return err
+		}
+		if resp.Err != nil && resp.Err.Code != enginepb.ErrorCode_None {
```

Comment:
In which scenario will the rpc call returns `enginepb.ErrorCode_None` error. And why doesn't `Heartbeat` or `QueryMetaStore` handle such error?

Reply:
Handling `enginepb.ErrorCode_None` is to safeguard against this case. This case probably never happens.
`Heartbeat`'s error handling will be refactored in future works. `Heartbeat`'s error handling strategy is unique, so we are not dealing with it for now.